From 72d7817f7034b160e2c9a9fb149f6aec2e265ffd Mon Sep 17 00:00:00 2001
From: "zhijie.lv" <zhijie.lv@baifendian.com>
Date: Tue, 1 Mar 2016 20:45:25 +0800
Subject: [PATCH] resolve the rebalance error

---
 cmd/cconfig/rebalancer.go | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cmd/cconfig/rebalancer.go b/cmd/cconfig/rebalancer.go
index 6d26e1c..76ada85 100644
--- a/cmd/cconfig/rebalancer.go
+++ b/cmd/cconfig/rebalancer.go
@@ -114,7 +114,7 @@ func Rebalance() error {
 	for _, node := range livingNodes {
 		for len(node.CurSlots) > targetQuota[node.GroupId] {
 			for _, dest := range livingNodes {
-				if dest.GroupId != node.GroupId && len(dest.CurSlots) < targetQuota[dest.GroupId] {
+				if dest.GroupId != node.GroupId && len(dest.CurSlots) < targetQuota[dest.GroupId] && len(node.CurSlots) > targetQuota[node.GroupId] {
 					slot := node.CurSlots[len(node.CurSlots)-1]
 					// create a migration task
 					info := &MigrateTaskInfo{
-- 
1.9.5.msysgit.1

