From 2b0f03e4f219a25b06c006e3ee3a4c8da0045d19 Mon Sep 17 00:00:00 2001
From: xuzhou <xuzhou1@jd.com>
Date: Thu, 22 Jun 2017 11:06:40 +0800
Subject: [PATCH 18/39] Optimize set command with ex/px when updating aof.

---
 src/aof.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/aof.c b/src/aof.c
index 4a179bd6..ea9ac47a 100644
--- a/src/aof.c
+++ b/src/aof.c
@@ -536,16 +536,16 @@ void feedAppendOnlyFile(struct redisCommand *cmd, int dictid, robj **argv, int a
         buf = catAppendOnlyGenericCommand(buf,3,tmpargv);
         decrRefCount(tmpargv[0]);
         buf = catAppendOnlyExpireAtCommand(buf,cmd,argv[1],argv[2]);
-    } else if (cmd->proc == setCommand) {
+    } else if (cmd->proc == setCommand && argc > 3) {
         int i;
         robj *exarg = NULL, *pxarg = NULL;
         /* Translate SET [EX seconds][PX milliseconds] to SET and PEXPIREAT */
         buf = catAppendOnlyGenericCommand(buf,3,argv);
         for (i = 3; i < argc; i ++) {
-            if (sdsEncodedObject(argv[i]) && !strcasecmp(argv[i]->ptr, "ex"))
+            if (!strcasecmp(argv[i]->ptr, "ex"))
                 exarg = argv[i+1];
 
-            if (sdsEncodedObject(argv[i]) && !strcasecmp(argv[i]->ptr, "px"))
+            if (!strcasecmp(argv[i]->ptr, "px"))
                 pxarg = argv[i+1];
         }
         serverAssert(!(exarg && pxarg));
-- 
2.14.1

