From 61c78a5215d9e97ea7c743bc6ee9dab8b634068e Mon Sep 17 00:00:00 2001
From: antirez <antirez@gmail.com>
Date: Fri, 19 May 2017 11:10:36 +0200
Subject: [PATCH 20/39] Collect fork() timing info only if fork succeeded.

---
 src/rdb.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/rdb.c b/src/rdb.c
index da660671..5b4a6cc2 100644
--- a/src/rdb.c
+++ b/src/rdb.c
@@ -1690,9 +1690,6 @@ int rdbSaveToSlavesSockets(void) {
         exitFromChild((retval == C_OK) ? 0 : 1);
     } else {
         /* Parent */
-        server.stat_fork_time = ustime()-start;
-        server.stat_fork_rate = (double) zmalloc_used_memory() * 1000000 / server.stat_fork_time / (1024*1024*1024); /* GB per second. */
-        latencyAddSampleIfNeeded("fork",server.stat_fork_time/1000);
         if (childpid == -1) {
             serverLog(LL_WARNING,"Can't save in background: fork: %s",
                 strerror(errno));
@@ -1715,6 +1712,10 @@ int rdbSaveToSlavesSockets(void) {
             close(pipefds[0]);
             close(pipefds[1]);
         } else {
+            server.stat_fork_time = ustime()-start;
+            server.stat_fork_rate = (double) zmalloc_used_memory() * 1000000 / server.stat_fork_time / (1024*1024*1024); /* GB per second. */
+            latencyAddSampleIfNeeded("fork",server.stat_fork_time/1000);
+
             serverLog(LL_NOTICE,"Background RDB transfer started by pid %d",
                 childpid);
             server.rdb_save_time_start = time(NULL);
-- 
2.14.1

