From 97d50344c4f856d024053eafded74b86d77504e8 Mon Sep 17 00:00:00 2001
From: Leon Chen <chen.bao.yi@qq.com>
Date: Tue, 18 Jul 2017 02:28:24 -0500
Subject: [PATCH 31/39] fix mismatch argument

---
 src/cluster.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/cluster.c b/src/cluster.c
index ea9b8c0a..f8321e2c 100644
--- a/src/cluster.c
+++ b/src/cluster.c
@@ -4007,7 +4007,7 @@ void clusterCommand(client *c) {
             }
             if ((n = clusterLookupNode(c->argv[4]->ptr)) == NULL) {
                 addReplyErrorFormat(c,"I don't know about node %s",
-                    (char*)c->argv[3]->ptr);
+                    (char*)c->argv[4]->ptr);
                 return;
             }
             server.cluster->importing_slots_from[slot] = n;
-- 
2.14.1

