From f59b4b93d9211c9888ab044caf2038fa6886424f Mon Sep 17 00:00:00 2001
From: antirez <antirez@gmail.com>
Date: Tue, 11 Apr 2017 16:54:27 +0200
Subject: [PATCH 09/19] Fix preprocessor if/else chain broken in order to fix
 #3927.

---
 src/zmalloc.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/zmalloc.c b/src/zmalloc.c
index d795f448..edfbd570 100644
--- a/src/zmalloc.c
+++ b/src/zmalloc.c
@@ -414,6 +414,9 @@ size_t zmalloc_get_memory_size(void) {
         return (size_t)size;
     return 0L;          /* Failed? */
 #else
+    return 0L;          /* Unknown method to get the data. */
+#endif
+#else
     return 0L;          /* Unknown OS. */
 #endif
 }
-- 
2.13.1

