From 8da9a167e08c6cb0ec1e789f322f47ff77128295 Mon Sep 17 00:00:00 2001
From: xuchengxuan <xuchengxuan@xiaomi.com>
Date: Wed, 14 Jun 2017 16:42:21 +0800
Subject: [PATCH 14/19] Fixed comments of slowlog duration

---
 src/slowlog.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/slowlog.h b/src/slowlog.h
index 81df0b05..6b8740e8 100644
--- a/src/slowlog.h
+++ b/src/slowlog.h
@@ -35,7 +35,7 @@ typedef struct slowlogEntry {
     robj **argv;
     int argc;
     long long id;       /* Unique entry identifier. */
-    long long duration; /* Time spent by the query, in nanoseconds. */
+    long long duration; /* Time spent by the query, in microseconds. */
     time_t time;        /* Unix time at which the query was executed. */
 } slowlogEntry;
 
-- 
2.13.1

